### PR TITLE
Added some base OneLogin rules

### PR DESCRIPTION
**Pull request by nhakmiller**
### Background

Added two base OneLogin rules.

### Changes

* Added a rule to detect unusual OneLogin authentications
* Added a rule to detect the assignment of user permissions

### Testing

* Unit tests

### CODE REVIEW

File: onelogin_rules/onelogin_unusual_login.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    # Skip API call if this is a unit test
+    if __name__ == 'PolicyApiTestingPolicy':
+        resp = lambda: None
+        setattr(resp, 'status_code', 200)
+        setattr(resp, 'text', event['api_data'])
+    else:
+        resp = requests.get(url)
+
+    if resp.status_code != 200:
+        # Could raise an exception here for ops team to look into
```
Comment (austinbyers) [inline onelogin_rules/onelogin_unusual_login.py:23]: Should we print or log a warning, then?

File: onelogin_rules/onelogin_unusual_login.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+def rule(event):
+    # Pre-filter to save compute time where possible. event_type_id = 5 is login events.
+    if event.get('event_type_id') != 5 or event.get('ipaddr') is None:
+        return False
+
+    # Lookup geo-ip data via API call
+    url = 'https://ipinfo.io/' + event['ipaddr'] + '/geo'
+
+    # Skip API call if this is a unit test
+    if __name__ == 'PolicyApiTestingPolicy':
```
Comment (austinbyers) [inline onelogin_rules/onelogin_unusual_login.py:15]: Whatever works! :)

File: onelogin_rules/onelogin_unusual_login.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        'city', '<CITY>')
+
+    # Lookup & store persistent data
+    event_key = get_key(event)
+    last_login_info = get_string_set(event_key)
+    if not last_login_info:
+        # Store this as the first login if we've never seen this user login before
+        put_string_set(event_key, [json.dumps({login_tuple: 1})])
+        return False
+    last_login_info = json.loads(last_login_info.pop())
```
Comment (austinbyers) [inline onelogin_rules/onelogin_unusual_login.py:36]: I'm having a hard time picturing what this is. A set of region strings? Perhaps a comment with an example response from the API would be helpful

File: onelogin_rules/onelogin_unusual_login.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    last_login_info[login_tuple] = last_login_info.get(login_tuple, 0) + 1
+    put_string_set(event_key, [json.dumps(last_login_info)])
+
+    tuple_count = last_login_info[login_tuple]
+    higher_tuples = 0
+    for tcount in last_login_info.values():
+        if tcount > tuple_count:
+            higher_tuples += 1
+        if higher_tuples == 3:
+            return True
```
Comment (austinbyers) [inline onelogin_rules/onelogin_unusual_login.py:47]: So what does this mean? We alert if the number of "unusual" logins is greater than a threshold? How do we determine what's unusual, or is that done by the API?

File: onelogin_rules/onelogin_unusual_login.py (hunk 2 above)
Comment (nhakmiller) [reply onelogin_rules/onelogin_unusual_login.py:15]: I found out after writing this that it only works in the actual rules engine, not with the panther_analysis_tool :/ All the more reason to switch to the Panther public API when it's ready!

File: onelogin_rules/onelogin_unusual_login.py (hunk 1 above)
Comment (nhakmiller) [reply onelogin_rules/onelogin_unusual_login.py:23]: Hmm good suggestion, I know everything that gets printed in the rules engine gets output to CloudWatch, but I'm not sure what formatting would be needed for our CloudWatch alarms to pick up on it. My main thought for this comment was that once we have better error reporting on the rules engine (i.e. straight to users, not to ops team) then we could raise an exception here. I will look into it some more.

File: onelogin_rules/onelogin_unusual_login.py (hunk 3 above)
Comment (nhakmiller) [reply onelogin_rules/onelogin_unusual_login.py:36]: Ah good point, a lot of the context was initially discussed verbally while working on this rule. I'll add some comments to transfer that context into code.

Comment (nhakmiller) [reply onelogin_rules/onelogin_unusual_login.py:42]: I added a comment about what the rule is trying to accomplish.

File: onelogin_rules/onelogin_unusual_login.py (hunk 4 above)
Comment (nhakmiller) [reply onelogin_rules/onelogin_unusual_login.py:55]: I added a comment here as well explaining the logic a bit. We could probably do some fancy python list comprehension to make this a one liner if we want, but that probably wouldn't make the logic anymore clear.

**Pull request merged**